runtime/compiler/optimizer/J9ValuePropagation.cpp

```diff
@@ -2188,20 +2188,38 @@ J9::ValuePropagation::innerConstrainAcall(TR::Node *node)
          static char *enableDynamicObjectClone = feGetEnv("TR_enableDynamicObjectClone");
          // Dynamic cloning kicks in when we attempt to make direct call to Object.clone
          // or J9VMInternals.primitiveClone where the cloned object is an array.
+
+         // Vector intrinsics are defined in VectorSupport.java. They all take object of type
+         // Class as a  parameter which represents the class of the vectors
+         // they are invoked on e.g. Float128Vector.class. That contraint needs to be propagated
+         // to their return type.
+         // load() has it as a first parameter and binaryOp as a second(first is opcode)
+         bool vectorSupportIntrinsic =
+            method->getRecognizedMethod() == TR::jdk_internal_vm_vector_VectorSupport_load ||
+            method->getRecognizedMethod() == TR::jdk_internal_vm_vector_VectorSupport_binaryOp;
+         
          if (method->getRecognizedMethod() == TR::java_lang_Object_clone
-             || method->getRecognizedMethod() == TR::java_lang_J9VMInternals_primitiveClone)
+             || method->getRecognizedMethod() == TR::java_lang_J9VMInternals_primitiveClone
+             || vectorSupportIntrinsic)
             {
             bool isGlobal;
-            if (method->getRecognizedMethod() == TR::java_lang_Object_clone)
+            if (method->getRecognizedMethod() == TR::jdk_internal_vm_vector_VectorSupport_binaryOp)
+               constraint = getConstraint(node->getSecondChild(), isGlobal);
+            else if (method->getRecognizedMethod() == TR::java_lang_Object_clone ||
+                vectorSupportIntrinsic)
               constraint = getConstraint(node->getFirstChild(), isGlobal);
             else
               constraint = getConstraint(node->getLastChild(), isGlobal);
 
+            if (vectorSupportIntrinsic)
```

Comment:
Is it intentional that this does not check for whether tracing of the VP optimization is enabled or not in addition to the above condition ? i.e. `trace()`

Reply:
No, it's by mistake. Thanks for noticing.

Reply:
Added trace() check.